File: packages/go/first.go
// Package arrayfirst returns the first element or first n elements of a slice.
// Go port of https://github.com/mattstrick/array-first-javascript
package arrayfirst

// First returns the first element of slice, and false if the slice is empty.
func First[T any](slice []T) (T, bool) {
	var zero T
	if len(slice) == 0 {
		return zero, false
	}
	return slice[0], true
}

// FirstN returns the first n elements. If n < 1 it is treated as 1.
// When n is 1, returns a single-element slice.
func FirstN[T any](slice []T, n int) ([]T, bool) {
	if len(slice) == 0 {
		return nil, false
	}
	if n < 1 {
		n = 1
	}
	if n > len(slice) {
		n = len(slice)
	}
	return slice[:n], true
}

File: packages/csharp/First.cs
namespace ArrayFirst;

/// <summary>
/// Get the first element or first n elements of a list.
/// C# port of https://github.com/mattstrick/array-first-javascript
/// </summary>
public static class First
{
    /// <summary>
    /// Returns the first element, or null if the list is empty.
    /// </summary>
    public static T? ArrayFirst<T>(IList<T>? arr)
    {
        if (arr == null)
            throw new ArgumentException("array-first expects a list as the first argument.");
        if (arr.Count == 0)
            return default;
        return arr[0];
    }

    /// <summary>
    /// Returns the first element if n is 1, otherwise the first n elements.
    /// </summary>
    public static object? ArrayFirst<T>(IList<T>? arr, int n)
    {
        if (arr == null)
            throw new ArgumentException("array-first expects a list as the first argument.");
        if (arr.Count == 0)
            return null;
        var take = n < 1 ? 1 : n;
        var slice = arr.Take(take).ToList();
        if (take == 1)
            return slice[0];
        return slice;
    }
}
